### PR TITLE
Regroup CI testing

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Part of https://github.com/neondatabase/neon/pull/2410 and https://github.com/neondatabase/neon/pull/2407

* adds `hashFiles('rust-toolchain.toml')`  into Rust cache keys, thus removing one of the manual steps to do when upgrading rustc
* copies Python and Rust style checks from the `codestyle.yml` workflow
* adjusts shell defaults in the main workflow
* replaces `codestyle.yml` with a `neon_extra_builds.yml` worlflow

The new workflow runs on commits to `main` (`codestyle.yml` was run per PR), and runs two custom builds on GH agents:

* macos-latest, to ensure the entire project compiles on it (no tests run)

There were no frequent breakages on macOs in our builds, so we can check it rarely without making every storage PR to wait for it to complete.
The updated mac build use release builds now, so presumably should work a bit faster due to overall smaller files to cache between builds.

* ubuntu-latest, without caches, to produce full compilation stats for Rust builds and upload it as an artifact to GitHub

Old `clippy build --timings` stats were collected from the builds that use caches and incremental calculation hence never could produce a full report, it got removed.

### CODE REVIEW

File: .github/workflows/neon_extra_builds.yml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        run: ./run_clippy.sh
+
+  gather-rust-build-stats:
+    timeout-minutes: 90
+    runs-on: ubuntu-latest
+
+    env:
+      BUILD_TYPE: release
+      # build with incremental compilation produce partial results
+      # so do not attempt to cache this build, also disable the imcremental compilation
```
Comment (koivunej) [inline .github/workflows/neon_extra_builds.yml:94]:
```suggestion
      # so do not attempt to cache this build, also disable the incremental compilation
```

**Review by koivunej — APPROVED** [review]
I think this is looking good. Have a question about the custom caching strat even on github runners.

File: .github/workflows/neon_extra_builds.yml [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+      - name: Cache cargo deps
+        uses: actions/cache@v3
+        with:
+          path: |
+            ~/.cargo/registry
+            !~/.cargo/registry/src
+            ~/.cargo/git
+            target
+          key: v1-${{ runner.os }}-cargo-${{ hashFiles('./Cargo.lock') }}-${{ hashFiles('./rust-toolchain.toml') }}-rust
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply .github/workflows/neon_extra_builds.yml:79]: I like the runner and once wanted to add it too, actually.
My main question now is what kind of storage does it use for caching? If it essentially reuses `actions/cache@v3` one (as we do now too), we have a way to improve here and cache such things ourselves with faster S3 `./.github/actions/upload` task we use.

No strong objections against using that plugin too, just feels like we're not going to change even the key for a while with the new model now 🙂

**Comment by SomeoneToIgnore** [comment]
Extra builds run example: https://github.com/neondatabase/neon/actions/runs/3658677345

**Review by hlinnaka — APPROVED** [review]

File: .github/workflows/neon_extra_builds.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply .github/workflows/neon_extra_builds.yml:76]: Oki I didn't know about the upcoming s3 cache usage. I think it does attempt some cleaning up of the cache compared to this, in general just trying to go with the mainstream solution if we are now changing these to mainstream runners. Have you tried it?

[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply .github/workflows/neon_extra_builds.yml:76]: Have not tried this, just checked out the yaml declaration options and the code a bit.
Yeah, I would say, I'd approve the PR that brings this in, but somehow don't feel sold to the idea to do that myself.

I think I'm just being lazy here and anticipating more changes to land on top of these workflows.
I will resolve the discussion, but thanks for bringing it up: if I find myself discussing the same thing next time, I'm definitely creating a PR 🙂

**Pull request merged**